flow/flow.h

```diff
@@ -235,6 +235,109 @@ struct union_like_traits<ErrorOr<T>> : std::true_type {
 	}
 };
 
+template <class T>
+class CachedSerialization {
+public:
+	constexpr static FileIdentifier file_identifier = FileIdentifierFor<T>::value;
+
+	enum class SerializeType { None, Binary, Object };
+
+	CachedSerialization() : cacheType(SerializeType::None) {}
+	explicit CachedSerialization(const T& data) : data(data), cacheType(SerializeType::None) {}
+	
+	const T& read() const { return data; }
+
+	T& mutate() {
+		cacheType = SerializeType::None;
+		return data;
+	}
+
+	//This should only be called from the ObjectSerializer load function
+	Standalone<StringRef> getCache() const {
+		if(cacheType != SerializeType::Object) {
+			cache = ObjectWriter::toValue(ErrorOr<EnsureTable<T>>(data), AssumeVersion(currentProtocolVersion));
+			cacheType = SerializeType::Object;
+		}
+		return cache;
+	}
+
+	bool operator == (CachedSerialization<T> const& rhs) const {
+		return data == rhs.data;
+	}
+	bool operator != (CachedSerialization<T> const& rhs) const {
+		return !(*this == rhs);
+	}
+	bool operator < (CachedSerialization<T> const& rhs) const {
+		return data < rhs.data;
+	}
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		if(Ar::isDeserializing) { 
+			cache = Standalone<StringRef>();
+			cacheType = SerializeType::None;
+			serializer(ar, data);
+		} else {
+			if(cacheType != SerializeType::Binary) {
+				cache = BinaryWriter::toValue(data, AssumeVersion(currentProtocolVersion));
+				cacheType = SerializeType::Binary;
+			}
+			ar.serializeBytes( const_cast<uint8_t *>(cache.begin()), cache.size() );
+		}
+	}
+
+private:
+	T data;
+	mutable SerializeType cacheType;
+	mutable Standalone<StringRef> cache; 
+};
+
+// this special case is needed - the code expects
+// Standalone<T> and T to be equivalent for serialization
+namespace detail {
+
+template <class T, class Context>
+struct LoadSaveHelper<CachedSerialization<T>, Context> : Context {
+	LoadSaveHelper(const Context& context)
+		: Context(context), helper(context) {}
+
+	void load(CachedSerialization<T>& member, const uint8_t* current) {
+		helper.load(member.mutate(), current);
+	}
+
+	template <class Writer>
+	RelativeOffset save(const CachedSerialization<T>& member, Writer& writer, const VTableSet* vtables) {
+		throw internal_error();
+	}
+
+private:
+	LoadSaveHelper<T, Context> helper;
+};
+
+} // namespace detail
+
+template <class V>
+struct serialize_raw<ErrorOr<EnsureTable<CachedSerialization<V>>>> : std::true_type {
+	template <class Context>
+	static uint8_t* save_raw(Context& context, const ErrorOr<EnsureTable<CachedSerialization<V>>>& obj) { 
+		auto cache = obj.present() ? obj.get().getCache() : ObjectWriter::toValue(ErrorOr<EnsureTable<V>>(obj.getError()), AssumeVersion(currentProtocolVersion));
+		uint8_t* out = context.allocate(cache.size());
+		memcpy(out, cache.begin(), cache.size());
+		return out;
+	}
+};
+
+template<class V>
+struct scalar_traits<CachedSerialization<V>> : std::true_type {
```

Comment:
I think one problem is that the vtable for `V` won't get collected, since flatbuffers thinks it doesn't have a vtable

Reply:
I think that should be fine though, since the save path for ErrorOr<EnsureTable<CachedSerialization<V>>> never gets called